app/classes/visual_group_data.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+class VisualGroupData
+  attr_accessor :query
+
+  def initialize(name, vote_limit)
+    self.query = tables[:observation_images]
+    add_joins
+    add_project
+    add_conditions(name, vote_limit)
+    query.order(tables[:observations][:vote_cache].desc)
+  end
+
+  def sql_query
+    self.query.to_sql
+  end
+
+  private
+
+  def tables
+    @tables ||= {
+      observation_images: ObservationImage.arel_table, # Arel::Table.new(:observation_images),
+      visual_group_images: VisualGroupImage.arel_table, # Arel::Table.new(:visual_group_images),
+      observations: Observation.arel_table,
+      names: Name.arel_table
+    }
+  end
+
+  def add_joins
+    join_table(:visual_group_images, :image_id,
+               attribute(:observation_images, :image_id),
+               Arel::Nodes::OuterJoin)
+    join_table(:observations, :id,
+               attribute(:observation_images, :observation_id))
+    join_table(:names, :id, attribute(:observations, :name_id))
+  end
+
+  def join_table(join_name, join_field, attribute, join_type = nil)
+    join_type ||= Arel::Nodes::InnerJoin
+    table = tables[join_name]
+    join_attribute = table[join_field]
+    self.query = query.join(table, join_type).on(join_attribute.eq(attribute))
+  end
+
+  def attribute(table_name, field)
+    tables[table_name][field]
+  end
+
+  def add_project
+    query.project(attribute(:observation_images, :image_id),
+                  attribute(:visual_group_images, :included),
+                  attribute(:observations, :vote_cache))
+  end
+
+  def add_conditions(name, vote_limit)
+    query.where(tables[:names][:text_name].matches("%#{name}%"))
+    query.where(tables[:observations][:vote_cache].gteq(vote_limit))
+  end
+end
```

Comment:
The only potential for using model scopes here that i can spot, is maybe instead of joining to `Name.arel_table` with the conditions, you define `tables[:names]` as `Name.text_name_includes(name)`... passing the `name` var down to `def tables`. You could do the same thing for Observations, but you'd have to add a new scope like `Observation.votes_above(limit)`.

I don't know if that's usable, or what that gets you, but it's nice that scopes are defined in one place — so if the definition of a scope needs to change, it changes everywhere.

Reply:
Hmm, tried it, but couldn't get it to fly.  It complained about "no implicit conversion of Symbol into Integer" when it did the `join_table` with :names after I make these changes:
```% git diff
diff --git a/app/classes/visual_group_data.rb b/app/classes/visual_group_data.rb
index 8fe0e2635..bdc9ee5a5 100644
--- a/app/classes/visual_group_data.rb
+++ b/app/classes/visual_group_data.rb
@@ -1,9 +1,11 @@
 # frozen_string_literal: true
 
 class VisualGroupData
+  attr_accessor :name
   attr_accessor :query
 
   def initialize(name, vote_limit)
+    self.name = name
     self.query = tables[:observation_images]
     add_joins
     add_project
@@ -22,7 +24,7 @@ class VisualGroupData
       observation_images: ObservationImage.arel_table,
       visual_group_images: VisualGroupImage.arel_table,
       observations: Observation.arel_table,
-      names: Name.arel_table
+      names: Name.text_name_includes(name) # Name.arel_table
     }
   end
 
@@ -53,7 +55,7 @@ class VisualGroupData
   end
 
   def add_conditions(name, vote_limit)
-    query.where(tables[:names][:text_name].matches("%#{name}%"))
+    # query.where(tables[:names][:text_name].matches("%#{name}%"))
     query.where(tables[:observations][:vote_cache].gteq(vote_limit))
   end
 end
```

Reply:
Aha. Sorry for not testing that first.
🐥 This guy is telling me the place for the scope might be rather in the conditions, like

```ruby
   def add_conditions(name, vote_limit)
     query.where(Name.text_name_includes(name))
     query.where(tables[:observations][:vote_cache].gteq(vote_limit))
   end
```

I'm standing by to help Jason with the email update. Can test later, or you can try locally.